p2panda-rs/src/storage_provider/traits/entry_store.rs

```diff
@@ -18,29 +18,136 @@ use crate::storage_provider::traits::AsStorageEntry;
+    /// successfully the skiplink entry is returned.
+    async fn try_get_skiplink(
+        &self,
+        entry: &StorageEntry,
+    ) -> Result<Option<StorageEntry>, EntryStorageError> {
+        // If a skiplink isn't required and it wasn't provided, return already now
+        if !is_lipmaa_required(entry.seq_num().as_u64()) && entry.skiplink_hash().is_none() {
+            return Ok(None);
+        };
+
+        // Derive the expected skiplink seq number from this entries sequence number
+        let expected_skiplink = match entry.seq_num().skiplink_seq_num() {
+            // Retrieve the expected skiplink from the database
+            Some(seq_num) => {
+                let expected_skiplink_entry = self
+                    .entry_at_seq_num(&entry.author(), &entry.log_id(), &seq_num)
+                    .await?
+                    .ok_or_else(|| EntryStorageError::ExpectedSkiplinkMissing(entry.hash()))?;
+                Some(expected_skiplink_entry)
+            }
+            // Or if there is no skiplink for entries at this sequence number return None
+            None => None,
+        };
+
+        // compare the expected skiplink hash and the stated skiplink hash
+        if expected_skiplink.clone().map(|entry| entry.hash()) != entry.skiplink_hash() {
+            return Err(EntryStorageError::InvalidSkiplinkPassed(entry.hash()));
+        }
+        Ok(expected_skiplink)
+    }
+    /// Get the latest Bamboo entry of an author's log.
+    ///
+    /// Returns a result containing an entry wrapped in an option. If no log was
+    /// could be found at this author - log location then None is returned.
+    /// Errors when a fatal storage error occurs.
     async fn latest_entry(
         &self,
         author: &Author,
         log_id: &LogId,
     ) -> Result<Option<StorageEntry>, EntryStorageError>;
 
-    /// Return vector of all entries of a given schema.
+    /// Get a vector of all entries of a given schema.
+    ///
+    /// Returns a result containing vector of entries wrapped in an option.
+    /// If no schema with this id could be found then None is returned.
+    /// Errors when a fatal storage error occurs.
     async fn by_schema(&self, schema: &SchemaId) -> Result<Vec<StorageEntry>, EntryStorageError>;
 
-    /// Determine skiplink entry hash ("lipmaa"-link) for entry in this log, return `None` when no
-    /// skiplink is required for the next entry.
-    async fn determine_skiplink(
+    /// Get all entries of a log from a specified sequence number up to passed max number of entries.
+    ///
+    /// Returns a vector of entries the length of which will not be greater than the max number
+    /// passed into the method. Fewer may be returned if the end of the log is reached.
+    async fn get_paginated_log_entries(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+        seq_num: &SeqNum,
+        max_number_of_entries: usize,
```

Comment:
I think `page_size` would make sense here to go with the idea of pagination.

Reply:
I really like `get_certificate_pool()` changed that now. Not sure about `page_size`, but maybe I'm not familiar with paginated term conventions?